lib/vwo/core/variation_decider.rb

```diff
@@ -226,6 +275,108 @@ def get_variation_of_campaign_for_user(user_id, campaign)
 
       private
 
+      # Evaluate all the variations in the campaign to find
+      #
+      # @param[String]  :user_id      The unique key assigned to User
+      # @param[Hash]    :campaign     Campaign hash for Unique campaign key
+      #
+      # @return[Hash]
+
+      def evaluate_whitelisting(user_id, campaign, api_name, campaign_key, variation_targeting_variables = {})
+        if variation_targeting_variables.nil?
+          variation_targeting_variables = { 'vwo_user_id' => user_id }
+        else
+          variation_targeting_variables['vwo_user_id'] = user_id
+        end
+        targeted_variations = []
+
+        campaign['variations'].each do |variation|
+          segments = get_segments(variation)
+          is_valid_segments = valid_value?(segments)
+          if is_valid_segments
+            if @segment_evaluator.evaluate(campaign_key, user_id, segments, variation_targeting_variables)
+              targeted_variations.push(variation)
+              status = StatusEnum::PASSED
+            else
+              status = StatusEnum::FAILED
+            end
+            @logger.log(
+              LogLevelEnum::DEBUG,
+              format(
+                LogMessageEnum::InfoMessages::SEGMENTATION_STATUS,
+                file: FILE,
+                campaign_key: campaign_key,
+                user_id: user_id,
+                status: status,
+                custom_variables: variation_targeting_variables,
+                variation_name: variation['name'],
+                segmentation_type: SegmentationTypeEnum::WHITELISTING,
+                api_name: api_name
+              )
+            )
+          else
+            @logger.log(
+              LogLevelEnum::DEBUG,
+              format(
+                LogMessageEnum::InfoMessages::SKIPPING_SEGMENTATION,
```

Comment:
shouldn't be whitelisting failed?

Reply:
no , this part account for segmentation of all the variants , log is for internal segmentation per variation. if a segment is not valid or empty. then only that segment get skipped and it will evaluate next segment.